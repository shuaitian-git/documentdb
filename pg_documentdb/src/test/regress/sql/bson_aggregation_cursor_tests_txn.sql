SET search_path TO documentdb_api_catalog, documentdb_api, documentdb_core, public;
SET documentdb.next_collection_id TO 3400;
SET documentdb.next_collection_index_id TO 3400;

set documentdb.enablePrimaryKeyCursorScan to on;

-- insert 100 documents
DO $$
DECLARE i int;
BEGIN
FOR i IN 1..10 LOOP
PERFORM documentdb_api.insert_one('db', 'aggregation_cursor_txn', FORMAT('{ "_id": %s, "sk": %s, "a": "%s", "c": [ %s "d" ] }',  i, mod(i, 2), repeat('Sample', 10), repeat('"' || repeat('a', 10) || '", ', 5))::documentdb_core.bson);
END LOOP;
END;
$$;

-- create an index to force non HOT path
SELECT documentdb_api_internal.create_indexes_non_concurrently('db', '{"createIndexes": "aggregation_cursor_txn", "indexes": [{"key": {"a": 1}, "name": "a_1" }]}', TRUE);

-- create a streaming cursor (that doesn't drain)
CREATE TEMP TABLE firstPageResponse AS
SELECT bson_dollar_project(cursorpage, '{ "cursor.firstBatch._id": 1, "cursor.id": 1 }'), continuation, persistconnection, cursorid FROM
    documentdb_api.find_cursor_first_page(database => 'db', commandSpec => '{ "find": "aggregation_cursor_txn", "batchSize": 5 }', cursorId => 4294967294);

SELECT * FROM firstPageResponse;

-- now drain it
SELECT continuation AS r1_continuation FROM firstPageResponse \gset
SELECT bson_dollar_project(cursorpage, '{ "cursor.nextBatch._id": 1, "cursor.id": 1 }'), continuation FROM
    documentdb_api.cursor_get_more(database => 'db', getMoreSpec => '{ "collection": "aggregation_cursor_txn", "getMore": 4294967294, "batchSize": 6 }', continuationSpec => :'r1_continuation');

-- repeat with an update
DROP TABLE firstPageResponse;

-- create a streaming cursor (that doesn't drain)
CREATE TEMP TABLE firstPageResponse AS
SELECT bson_dollar_project(cursorpage, '{ "cursor.firstBatch._id": 1, "cursor.id": 1 }'), continuation, persistconnection, cursorid FROM
    documentdb_api.find_cursor_first_page(database => 'db', commandSpec => '{ "find": "aggregation_cursor_txn", "batchSize": 5 }', cursorId => 4294967294);

SELECT * FROM firstPageResponse;

-- now update all the rows
SELECT documentdb_api.update('db', '{ "update": "aggregation_cursor_txn", "updates": [ { "q": { "_id": { "$gt": 0 } }, "u": { "$set": { "someField": "Updated" } }, "multi": true } ] }');

-- now drain it - should still drain.
SELECT continuation AS r1_continuation FROM firstPageResponse \gset
SELECT bson_dollar_project(cursorpage, '{ "cursor.nextBatch._id": 1, "cursor.id": 1 }'), continuation FROM
    documentdb_api.cursor_get_more(database => 'db', getMoreSpec => '{ "collection": "aggregation_cursor_txn", "getMore": 4294967294, "batchSize": 6 }', continuationSpec => :'r1_continuation');

SELECT documentdb_api.shard_collection('db', 'aggregation_cursor_txn', '{ "sk": "hashed" }', FALSE);


-- repeat with sharded setup
-- repeat with an update
DROP TABLE firstPageResponse;

-- create a streaming cursor (that doesn't drain)
CREATE TEMP TABLE firstPageResponse AS
SELECT bson_dollar_project(cursorpage, '{ "cursor.firstBatch._id": 1, "cursor.id": 1 }'), continuation, persistconnection, cursorid FROM
    documentdb_api.find_cursor_first_page(database => 'db', commandSpec => '{ "find": "aggregation_cursor_txn", "batchSize": 5 }', cursorId => 4294967294);

SELECT * FROM firstPageResponse;

-- now update all the rows
SELECT documentdb_api.update('db', '{ "update": "aggregation_cursor_txn", "updates": [ { "q": { "_id": { "$gt": 0 } }, "u": { "$set": { "someField": "Updated" } }, "multi": true } ] }');

-- now drain it - should still drain.
SELECT continuation AS r1_continuation FROM firstPageResponse \gset
SELECT bson_dollar_project(cursorpage, '{ "cursor.nextBatch._id": 1, "cursor.id": 1 }'), continuation FROM
    documentdb_api.cursor_get_more(database => 'db', getMoreSpec => '{ "collection": "aggregation_cursor_txn", "getMore": 4294967294, "batchSize": 6 }', continuationSpec => :'r1_continuation');
