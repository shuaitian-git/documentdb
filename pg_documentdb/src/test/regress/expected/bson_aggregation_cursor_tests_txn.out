SET search_path TO documentdb_api_catalog, documentdb_api, documentdb_core, public;
SET documentdb.next_collection_id TO 3400;
SET documentdb.next_collection_index_id TO 3400;
set documentdb.enablePrimaryKeyCursorScan to on;
-- insert 100 documents
DO $$
DECLARE i int;
BEGIN
FOR i IN 1..10 LOOP
PERFORM documentdb_api.insert_one('db', 'aggregation_cursor_txn', FORMAT('{ "_id": %s, "sk": %s, "a": "%s", "c": [ %s "d" ] }',  i, mod(i, 2), repeat('Sample', 10), repeat('"' || repeat('a', 10) || '", ', 5))::documentdb_core.bson);
END LOOP;
END;
$$;
NOTICE:  creating collection
-- create an index to force non HOT path
SELECT documentdb_api_internal.create_indexes_non_concurrently('db', '{"createIndexes": "aggregation_cursor_txn", "indexes": [{"key": {"a": 1}, "name": "a_1" }]}', TRUE);
                                                                                                   create_indexes_non_concurrently                                                                                                    
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 { "raw" : { "defaultShard" : { "numIndexesBefore" : { "$numberInt" : "1" }, "numIndexesAfter" : { "$numberInt" : "2" }, "createdCollectionAutomatically" : false, "ok" : { "$numberInt" : "1" } } }, "ok" : { "$numberInt" : "1" } }
(1 row)

-- create a streaming cursor (that doesn't drain)
CREATE TEMP TABLE firstPageResponse AS
SELECT bson_dollar_project(cursorpage, '{ "cursor.firstBatch._id": 1, "cursor.id": 1 }'), continuation, persistconnection, cursorid FROM
    documentdb_api.find_cursor_first_page(database => 'db', commandSpec => '{ "find": "aggregation_cursor_txn", "batchSize": 5 }', cursorId => 4294967294);
SELECT * FROM firstPageResponse;
                                                                                                                        bson_dollar_project                                                                                                                        |                                                                                                                                                                                                                                     continuation                                                                                                                                                                                                                                     | persistconnection |  cursorid  
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+------------
 { "cursor" : { "id" : { "$numberLong" : "4294967294" }, "firstBatch" : [ { "_id" : { "$numberInt" : "1" } }, { "_id" : { "$numberInt" : "2" } }, { "_id" : { "$numberInt" : "3" } }, { "_id" : { "$numberInt" : "4" } }, { "_id" : { "$numberInt" : "5" } } ] } } | { "qi" : { "$numberLong" : "4294967294" }, "qp" : false, "qk" : { "$numberInt" : "1" }, "qc" : { "find" : "aggregation_cursor_txn", "batchSize" : { "$numberInt" : "5" } }, "continuation" : [ { "table_name" : "documents_3400", "value" : { "$binary" : { "base64" : "AAAAAAUA", "subType" : "00" } }, "pk" : [ { "$numberLong" : "3400" }, { "" : { "$numberInt" : "5" } } ] } ], "numIters" : { "$numberInt" : "1" }, "sn" : NOW_SYS_VARIABLE } | f                 | 4294967294
(1 row)

-- now drain it
SELECT continuation AS r1_continuation FROM firstPageResponse \gset
SELECT bson_dollar_project(cursorpage, '{ "cursor.nextBatch._id": 1, "cursor.id": 1 }'), continuation FROM
    documentdb_api.cursor_get_more(database => 'db', getMoreSpec => '{ "collection": "aggregation_cursor_txn", "getMore": 4294967294, "batchSize": 6 }', continuationSpec => :'r1_continuation');
                                                                                                                   bson_dollar_project                                                                                                                    | continuation 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------
 { "cursor" : { "id" : { "$numberLong" : "0" }, "nextBatch" : [ { "_id" : { "$numberInt" : "6" } }, { "_id" : { "$numberInt" : "7" } }, { "_id" : { "$numberInt" : "8" } }, { "_id" : { "$numberInt" : "9" } }, { "_id" : { "$numberInt" : "10" } } ] } } | 
(1 row)

-- repeat with an update
DROP TABLE firstPageResponse;
-- create a streaming cursor (that doesn't drain)
CREATE TEMP TABLE firstPageResponse AS
SELECT bson_dollar_project(cursorpage, '{ "cursor.firstBatch._id": 1, "cursor.id": 1 }'), continuation, persistconnection, cursorid FROM
    documentdb_api.find_cursor_first_page(database => 'db', commandSpec => '{ "find": "aggregation_cursor_txn", "batchSize": 5 }', cursorId => 4294967294);
SELECT * FROM firstPageResponse;
                                                                                                                        bson_dollar_project                                                                                                                        |                                                                                                                                                                                                                                     continuation                                                                                                                                                                                                                                     | persistconnection |  cursorid  
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+------------
 { "cursor" : { "id" : { "$numberLong" : "4294967294" }, "firstBatch" : [ { "_id" : { "$numberInt" : "1" } }, { "_id" : { "$numberInt" : "2" } }, { "_id" : { "$numberInt" : "3" } }, { "_id" : { "$numberInt" : "4" } }, { "_id" : { "$numberInt" : "5" } } ] } } | { "qi" : { "$numberLong" : "4294967294" }, "qp" : false, "qk" : { "$numberInt" : "1" }, "qc" : { "find" : "aggregation_cursor_txn", "batchSize" : { "$numberInt" : "5" } }, "continuation" : [ { "table_name" : "documents_3400", "value" : { "$binary" : { "base64" : "AAAAAAUA", "subType" : "00" } }, "pk" : [ { "$numberLong" : "3400" }, { "" : { "$numberInt" : "5" } } ] } ], "numIters" : { "$numberInt" : "1" }, "sn" : NOW_SYS_VARIABLE } | f                 | 4294967294
(1 row)

-- now update all the rows
SELECT documentdb_api.update('db', '{ "update": "aggregation_cursor_txn", "updates": [ { "q": { "_id": { "$gt": 0 } }, "u": { "$set": { "someField": "Updated" } }, "multi": true } ] }');
                                                                update                                                                
--------------------------------------------------------------------------------------------------------------------------------------
 ("{ ""ok"" : { ""$numberDouble"" : ""1.0"" }, ""nModified"" : { ""$numberInt"" : ""10"" }, ""n"" : { ""$numberInt"" : ""10"" } }",t)
(1 row)

-- now drain it - should still drain.
SELECT continuation AS r1_continuation FROM firstPageResponse \gset
SELECT bson_dollar_project(cursorpage, '{ "cursor.nextBatch._id": 1, "cursor.id": 1 }'), continuation FROM
    documentdb_api.cursor_get_more(database => 'db', getMoreSpec => '{ "collection": "aggregation_cursor_txn", "getMore": 4294967294, "batchSize": 6 }', continuationSpec => :'r1_continuation');
                                                                                                                   bson_dollar_project                                                                                                                    | continuation 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------
 { "cursor" : { "id" : { "$numberLong" : "0" }, "nextBatch" : [ { "_id" : { "$numberInt" : "6" } }, { "_id" : { "$numberInt" : "7" } }, { "_id" : { "$numberInt" : "8" } }, { "_id" : { "$numberInt" : "9" } }, { "_id" : { "$numberInt" : "10" } } ] } } | 
(1 row)

SELECT documentdb_api.shard_collection('db', 'aggregation_cursor_txn', '{ "sk": "hashed" }', FALSE);
 shard_collection 
------------------
 
(1 row)

-- repeat with sharded setup
-- repeat with an update
DROP TABLE firstPageResponse;
-- create a streaming cursor (that doesn't drain)
CREATE TEMP TABLE firstPageResponse AS
SELECT bson_dollar_project(cursorpage, '{ "cursor.firstBatch._id": 1, "cursor.id": 1 }'), continuation, persistconnection, cursorid FROM
    documentdb_api.find_cursor_first_page(database => 'db', commandSpec => '{ "find": "aggregation_cursor_txn", "batchSize": 5 }', cursorId => 4294967294);
SELECT * FROM firstPageResponse;
                                                                                                                        bson_dollar_project                                                                                                                         |                                                                                                                                                                                                                                            continuation                                                                                                                                                                                                                                             | persistconnection |  cursorid  
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------+------------
 { "cursor" : { "id" : { "$numberLong" : "4294967294" }, "firstBatch" : [ { "_id" : { "$numberInt" : "2" } }, { "_id" : { "$numberInt" : "4" } }, { "_id" : { "$numberInt" : "6" } }, { "_id" : { "$numberInt" : "8" } }, { "_id" : { "$numberInt" : "10" } } ] } } | { "qi" : { "$numberLong" : "4294967294" }, "qp" : false, "qk" : { "$numberInt" : "1" }, "qc" : { "find" : "aggregation_cursor_txn", "batchSize" : { "$numberInt" : "5" } }, "continuation" : [ { "table_name" : "documents_3400", "value" : { "$binary" : { "base64" : "AAAAAAoA", "subType" : "00" } }, "pk" : [ { "$numberLong" : "174293393270269110" }, { "" : { "$numberInt" : "10" } } ] } ], "numIters" : { "$numberInt" : "1" }, "sn" : NOW_SYS_VARIABLE } | f                 | 4294967294
(1 row)

-- now update all the rows
SELECT documentdb_api.update('db', '{ "update": "aggregation_cursor_txn", "updates": [ { "q": { "_id": { "$gt": 0 } }, "u": { "$set": { "someField": "Updated" } }, "multi": true } ] }');
                                                               update                                                                
-------------------------------------------------------------------------------------------------------------------------------------
 ("{ ""ok"" : { ""$numberDouble"" : ""1.0"" }, ""nModified"" : { ""$numberInt"" : ""0"" }, ""n"" : { ""$numberInt"" : ""10"" } }",t)
(1 row)

-- now drain it - should still drain.
SELECT continuation AS r1_continuation FROM firstPageResponse \gset
SELECT bson_dollar_project(cursorpage, '{ "cursor.nextBatch._id": 1, "cursor.id": 1 }'), continuation FROM
    documentdb_api.cursor_get_more(database => 'db', getMoreSpec => '{ "collection": "aggregation_cursor_txn", "getMore": 4294967294, "batchSize": 6 }', continuationSpec => :'r1_continuation');
                                                                                                                   bson_dollar_project                                                                                                                   | continuation 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------
 { "cursor" : { "id" : { "$numberLong" : "0" }, "nextBatch" : [ { "_id" : { "$numberInt" : "1" } }, { "_id" : { "$numberInt" : "3" } }, { "_id" : { "$numberInt" : "5" } }, { "_id" : { "$numberInt" : "7" } }, { "_id" : { "$numberInt" : "9" } } ] } } | 
(1 row)

